lib/trino-filesystem-gcs/src/main/java/io/trino/filesystem/gcs/GcsInputFile.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.filesystem.gcs;
+
+import com.google.cloud.storage.Blob;
+import com.google.cloud.storage.Storage;
+import io.trino.filesystem.Location;
+import io.trino.filesystem.TrinoInput;
+import io.trino.filesystem.TrinoInputFile;
+import io.trino.filesystem.TrinoInputStream;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.time.Instant;
+import java.util.Optional;
+import java.util.OptionalLong;
+
+import static io.trino.filesystem.gcs.GcsUtils.getBlob;
+import static io.trino.filesystem.gcs.GcsUtils.handleGcsException;
+import static java.util.Objects.requireNonNull;
+
+public class GcsInputFile
+        implements TrinoInputFile
+{
+    private final GcsLocation location;
+    private final Storage storage;
+    private final int readBlockSize;
+    private OptionalLong length;
+    private OptionalLong predeclaredLength;
+    private Optional<Instant> lastModified = Optional.empty();
+
+    public GcsInputFile(GcsLocation location, Storage storage, int readBockSize, OptionalLong predeclaredLength)
+    {
+        this.location = requireNonNull(location, "location is null");
+        this.storage = requireNonNull(storage, "storage is null");
+        this.readBlockSize = readBockSize;
+        this.predeclaredLength = requireNonNull(predeclaredLength, "length is null");
+        this.length = OptionalLong.empty();
+    }
+
+    @Override
+    public TrinoInput newInput()
+            throws IOException
+    {
+        Blob blob = getBlob(storage, location);
```

Comment:
We shouldn't need to fetch metadata here. There's no API requirement to check that the file exists at this point.